platform/kobo/koreader.sh

```diff
@@ -36,14 +36,36 @@ if pkill -0 nickel ; then
 fi
 
 if [ "${FROM_NICKEL}" == "true" ] ; then
-	# Siphon a few things from nickel's env...
-	eval "$(xargs -n 1 -0 < /proc/$(pidof nickel)/environ | grep -e DBUS_SESSION_BUS_ADDRESS -e WIFI_MODULE -e PLATFORM -e WIFI_MODULE_PATH -e INTERFACE -e PRODUCT 2>/dev/null)"
-	export DBUS_SESSION_BUS_ADDRESS WIFI_MODULE PLATFORM WIFI_MODULE_PATH INTERFACE PRODUCT
+	# Detect if we were started from KFMon
+	if pkill -0 kfmon ; then
+		FROM_KFMON="true"
+	else
+		FROM_KFMON="false"
+	fi
+
+	if [ "${FROM_KFMON}" == "true" ] ; then
+		# Siphon nickel's full environment, since KFMon inherits such a minimal one...
+		for env in $(xargs -n 1 -0 < /proc/$(pidof nickel)/environ) ; do
+			export ${env}
+		done
+	else
+		# Siphon a few things from nickel's env...
+		eval "$(xargs -n 1 -0 < /proc/$(pidof nickel)/environ | grep -e DBUS_SESSION_BUS_ADDRESS -e WIFI_MODULE -e PLATFORM -e WIFI_MODULE_PATH -e INTERFACE -e PRODUCT 2>/dev/null)"
+		export DBUS_SESSION_BUS_ADDRESS WIFI_MODULE PLATFORM WIFI_MODULE_PATH INTERFACE PRODUCT
+	fi
 
 	# flush disks, might help avoid trashing nickel's DB...
 	sync
+	# Double the fun!
+	sleep 1
+	sync
 	# stop kobo software because it's running
-	killall nickel hindenburg fmon 2>/dev/null
+	killall nickel hindenburg sickel fickel fmon 2>/dev/null
+
+	# NOTE: Not particularly critical, we should be safe leaving it up, but since we reboot on exit anyway...
+	#if [ "${FROM_KFMON}" == "true" ] ; then
+	#	killall kfmon 2>/dev/null
+	#fi
```

Comment:
I was thinking of leaving it up for a while, to see if that unearthers KFMon bugs... ;p.

But in the end, yeah, it can go down (but only as soon as we've checked if we've been launched through it, which is basically around that block of code).

This makes me think that this check is kinda lame anyway, it triggers for everyone who has kfmon running, but not for instances when kfmon truly is our parent, which is what I should actually check!